### PR TITLE
Fix: individual /Resources directory are now properly produced for each document page

### PR DESCRIPTION
**Pull request by Lucas-C**
This change causes a slight increase of PDF documents, but makes their structure more valid regarding the PDF spec.

**Checklist**:

- [x] The GitHub pipeline is OK (green), meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [ ] A unit test is covering the code added / modified by this PR

- [ ] This PR is ready to be merged

- [ ] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [x] A mention of the change is present in `CHANGELOG.md`

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/py-pdf/fpdf2/blob/master/LICENSE).

### CODE REVIEW

**Comment by Lucas-C** [comment]
Could you please review this @andersonhc or @gmischler?

Despite the 377 reference test files modified, this PR onlys affects 2 source Python files and is relatively short.

**Comment by andersonhc** [comment]
I will try to make some time for a complete review by tomorrow.
For now I am running some tests and it's interesting your PR makes it clear we are letting unused fonts end up in the output document.

An example:
```python
from fpdf import FPDF

pdf = FPDF()
pdf.add_font("NotoSans", "B", "NotoSans-Bold.ttf")
pdf.add_font("NotoSans", "BI", "NotoSans-BoldItalic.ttf")
pdf.add_font("NotoSans", "I", "NotoSans-Italic.ttf")
pdf.add_font("NotoSans", "", "NotoSans-Regular.ttf")

pdf.set_font("NotoSans", "", 12)
pdf.single_resources_object = False

pdf.add_page()
pdf.multi_cell(w=pdf.epw, text="**Text in bold**", markdown=True)

pdf.add_page()
pdf.multi_cell(w=pdf.epw, text="__Text in italic__", markdown=True)

pdf.output("test1.pdf")
```

The page 1 will have "F1" and "F4" in the resources dictionary, and page 2 will have "F3" and "F4".
"F2" is added on the final document and not referenced at all. "F4" is added to the resource list because of `set_font()` although not used.

I have documents with many fallback fonts and there is a considerable amount of unused font data added to the documents.

Might be something to tackle in a future PR.

File: CHANGELOG.md [hunk 1]
```diff
@@ -25,6 +25,7 @@ This can also be enabled programmatically with `warnings.simplefilter('default',
 * [`FPDF.write_html()`](https://py-pdf.github.io/fpdf2/fpdf/fpdf.html#fpdf.fpdf.FPDF.write_html) now accepts a `tag_indents` parameter to control, for example, the indent of `<blockquote>` elements
 * allow to define custom `cell_fill_mode` logic for tables: [_Set cells background_ - documentation section](https://py-pdf.github.io/fpdf2/Tables.html#set-cells-background). Also added 2 new values: `TableCellFillMode.EVEN_ROWS` & `TableCellFillMode.EVEN_COLUMNS`: [documentation](https://py-pdf.github.io/fpdf2/fpdf/enums.html#fpdf.enums.TableCellFillMode)
 ### Fixed
+* individual `/Resources` directory are now properly produced for **each** document page. This causes a slight increase of PDF documents, but makes their structure more valid regarding the PDF spec. You can still use the old behaviour by setting `FPDF().single_resources_object = True`.
```
Comment (andersonhc) [inline CHANGELOG.md:28]: Suggestion to make it clearer the increase is the size of the documents:

* individual `/Resources` directories are now properly created for each document page. This change ensures better compliance with the PDF specification but results in a slight increase in the size of PDF documents. You can still use the old behavior by setting `FPDF().single_resources_object = True`.

**Review by andersonhc — APPROVED** [review]
I don't have anything else to add. The changes are simple and clear.
I didn't have time to visually inspect all the changed reference files but I did a few and everything looks OK.

File: CHANGELOG.md (hunk 1 above)
Comment (Lucas-C) [reply CHANGELOG.md:28]: Adopted!
Thank you 🙂

File: fpdf/output.py [hunk 2]
```diff
@@ -785,6 +779,40 @@ def _add_gfxstates(self):
...
+                    for (gfx_name, gfx_state) in gfxstate_objs_per_name.items()
+                    if gfx_name
+                    in self.fpdf.graphics_style_names_per_page_number[page_number]
+                }
+                page_obj.resources = self._add_resources_dict(
+                    page_font_objs_per_index,
+                    page_img_objs_per_index,
+                    page_gfxstate_objs_per_name,
+                )
+
```
[the remark this answers is not in the document]
Comment (Lucas-C) [reply fpdf/output.py:815]: Agreed!
Merging this "as it is" for now.

**Comment by Lucas-C** [comment]
> The page 1 will have "F1" and "F4" in the resources dictionary, and page 2 will have "F3" and "F4". "F2" is added on the final document and not referenced at all. "F4" is added to the resource list because of `set_font()` although not used.
> 
> I have documents with many fallback fonts and there is a considerable amount of unused font data added to the documents.
> 
> Might be something to tackle in a future PR.

I opened issue https://github.com/py-pdf/fpdf2/issues/1382 regarding this.

**Pull request merged**